terraform/transform_attach_schema.go

```diff
@@ -60,10 +62,13 @@ func (t *AttachSchemaTransformer) Transform(g *Graph) error {
 			mode := addr.Resource.Mode
 			typeName := addr.Resource.Type
 			providerAddr, _ := tv.ProvidedBy()
-			providerType := providerAddr.ProviderConfig.Type
 
-			// FIXME: where to lookup FQN from for real?
-			providerFqn := addrs.NewLegacyProvider(providerType).String()
+			var providerFqn string
+			if t.Config != nil {
+				providerFqn = t.Config.ProviderForLocalConfigAddr(providerAddr.ProviderConfig).String()
```

Comment:
(Continuing to review this commit-by-commit.)

I guess this is the point where things started to get tricky with `AbsProviderConfig` having the FQN inside? Because we've defined `node.ProvidedBy()` to return `addrs.AbsProviderConfig, bool` where the bool return value can be `false` to say "this provider config is actually a _local_ one, in spite of its type". :confounded: 

Looking at this problem again with fresh eyes today, I have a thought on a different way to solve this that might fit more nicely within the type system:

Since we've now renamed `addrs.ProviderConfig` to `addrs.LocalProviderConfig` anyway, could we define a new `addrs.ProviderConfig` that is a closed interface type whose concrete types are limited to the other two address types?

```go
// ProviderAddr represents either a local or absolute provider address, for
// situations where the decision of which to use is made at runtime rather
// than at compile time.
//
// LocalProviderAddr and AbsProviderAddr are the only two concrete types
// that a ProviderAddr value may have.
type ProviderAddr interface {
	providerAddr() // unexported so that no other package can implement this type
}
```

If we then make both `LocalProviderAddr` and `AbsProviderAddr` implement this interface, we could redefine `GraphNodeProviderConsumer` like this:

```go
type GraphNodeProviderConsumer interface {
	// ProvidedBy returns the address of the provider configuration the node
	// refers to. Can either return an exact addrs.AbsProviderConfig, if the
        // absolute address is already known, or an addrs.LocalProviderConfig
        // to indicate that resolution is necessary. If the result is
        // addrs.LocalProviderConfig, the a provider configuration from an
        // ancestor module may be selected instead.
	ProvidedBy() addrs.ProviderConfig
	// Set the resolved provider address for this resource.
	SetProvider(addrs.AbsProviderConfig)
}
```

A type switch on the `ProvidedBy` return value can then replace the weird `exact bool` return value to determine whether we're returning a local or an absolute address, depending on whether this graph node is representing something from the configuration or something from the state.

Finally, we could add a helper method to `configs.Config` to deal with resolving one of these local-or-absolute addresses into a provider FQN, since that seems to be the common operation we need for this transform and elsewhere:

```go
func (c *Config) ProviderForProviderConfig(addr addrs.ProviderConfig, module addrs.Module) (addrs.Provider, error) {
	// If addr is already absolute then just return its Provider
	// FQN value directly.
	// Otherwise, use the provider local name mapping tables
	// in the modules to figure out the provider FQN to return.
}
```

What do you think? If possible I'd like to let the Go type system help us, so we can use types to be clear with ourselves about what each value is representing. :thinking: